[cutlass backend] fix assertion that prevent self multiplication

henrylhtsang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #148233

# Problem:
In a matmul, sometimes some of the nodes are the same. Say `A @ A`. In that case, when writing the stride of node B, we have to figure out if we want lda or ldb, which points to the same node, and we have no way to differentiate which one.

# Solution
Just use whichever. Since they are the same.

# Question
What if we compile with `A @ A`, and then pass in `A @ B`? Well inductor guards will raise an error. 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

henrylhtsang committed:
[cutlass backend] remove assertion that prevent self multiplication

[ghstack-poisoned]

henrylhtsang committed:
Update on "[cutlass backend] remove assertion that prevent self multiplication"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

Skylion007 commented on torch/_inductor/codegen/cuda/cuda_kernel.py:
```diff
@@ -86,7 +86,6 @@ def find_layout_arg(
         matches = [
             arg for arg in self.layout_args.values() if arg.matches(node, attr, dim)
         ]
-        assert len(matches) <= 1, matches
```
Can we add an identity check here that if len(matches) >= 2 all items are the same?

henrylhtsang replied:
> Can we add an identity check here that if len(matches) >= 2 at least 2 of the items are the same?

they are not, interesting

```
[LayoutArg(node=StorageBox(
  InputBuffer(name='arg0_1', layout=FixedLayout('cuda:0', torch.float16, size=[128, 128], stride=[128, 1]))
), symbol='lda', attr='stride', dim=0), LayoutArg(node=StorageBox(
  InputBuffer(name='arg0_1', layout=FixedLayout('cuda:0', torch.float16, size=[128, 128], stride=[128, 1]))
), symbol='ldb', attr='stride', dim=0)]
```

henrylhtsang committed:
Update on "[cutlass backend] remove assertion that prevent self multiplication"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

henrylhtsang committed:
Update on "[cutlass backend] remove assertion that prevent self multiplication"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

henrylhtsang committed:
Update on "[cutlass backend] remove assertion that prevent self multiplication"


# Problem:
In a matmul, sometimes some of the nodes are the same. Say `A @ A`. In that case, when writing the stride of node B, we have to figure out if we want lda or ldb, which points to the same node, and we have no way to differentiate which one.

# Solution
Just use whichever. Since they are the same.

# Question
What if we compile with `A @ A`, and then pass in `A @ B`? Well inductor guards will raise an error. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

henrylhtsang commented:
@pytorchbot merge -i

henrylhtsang commented:
@pytorchbot revert -c nosignal -m "mistake in PR "

henrylhtsang commented:
@pytorchbot merge -i

henrylhtsang commented:
@pytorchbot revert -c nosignal -m "mistake in PR "


I don't know why my local changes didn't get pushed to here. My apologies

henrylhtsang committed:
Update on "[cutlass backend] fix assertion that prevent self multiplication "


# Problem:
In a matmul, sometimes some of the nodes are the same. Say `A @ A`. In that case, when writing the stride of node B, we have to figure out if we want lda or ldb, which points to the same node, and we have no way to differentiate which one.

# Solution
Just use whichever. Since they are the same.

# Question
What if we compile with `A @ A`, and then pass in `A @ B`? Well inductor guards will raise an error. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

henrylhtsang commented:
@pytorchbot merge -i
